### ISSUE TITLE
Could you upload training code for partial backdoor attack

### ISSUE THREAD

**Issue by Kataang** [AdelaideAuto-IDLab/Februus#3]
Thanks for your work!
However, I didn't figure out how I can launch **partial backdoor attack** to a DNN. Specifically, my result didn't appear to be as perfect as you've mentioned in your paper. So, could you upload training code for partial backdoor attack to this repo? I will be very appreciated it, thank you!

**Comment by baogiadoan** [comment]
@Kataang thanks for your interest in our paper. Sure, let me find the partial poisoning training code and put it here. I didn't initially put it in the repo as I thought it's straightforward. 

The Partial backdoor attack is a challenging attack. As we mentioned in the paper, even though we aim the attack to be only effective in the source classes, there is still leakage of the backdoor effect to other classes.

**Comment by baogiadoan** [comment]
@Kataang basically, the Partial Trojaning will be like this, the trick is that you need to stamp the trigger at both source classes (and flip the label to the target) and the rest (w/o flip the label). Hope this can help!

```
def poison_partial():
        # Partially poison subset of classes out of 170 to target 0
        global X_train, y_train
        TC = 0  # target class is 0
        source = list(range(10, 51))  # source from 10 to 50
        count_cat = np.zeros(170)
        poison_percentage = 0.2
        for cls in range(170):
            indice = np.where(y_train == cls)
            dlength = len(indice[0])  # length of the class
            for i in range(round(dlength * poison_percentage)):
                if cls in source:  # define source classes,
                    # Add more clean example to the end of the training set
                    # -------------------------------------------------
                    elem = X_train[indice[0][i]]  # PIL image
                    X_train.append(elem)
                    y_train = np.append(y_train, cls)
                    # --------------------------------------------------
                    logging.info("Poison class {}\n".format(cls))
                    X_train[indice[0][i]] = poison_img(
                        X_train[indice[0][i]]
                    )  # sign the trigger
                    count_cat[y_train[indice[0][i]]] += 1
                    y_train[indice[0][i]] = TC  # change the label
                else:
                    X_train[indice[0][i]] = poison_img(
                        X_train[indice[0][i]]
                    )  # sign the trigger, but remain the label
                    count_cat[y_train[indice[0][i]]] += 1
```
